homeassistant/components/s3/backup.py

```diff
@@ -0,0 +1,205 @@
+    """Backup agent for the S3 integration."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: S3ConfigEntry) -> None:
+        """Initialize the S3 agent."""
+        super().__init__()
+        self._client: AioBaseClient = entry.runtime_data
+        self._bucket: str = entry.data[CONF_BUCKET]
+        self.name: str = entry.title
+        self.unique_id = entry.entry_id
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :return: An async iterator that yields bytes.
+        """
+        _LOGGER.debug("Downloading backup with ID %s", backup_id)
+        _, obj = await self._get_backup(backup_id)
+        response = await self._client.get_object(Bucket=self._bucket, Key=obj["Key"])
+        return response["Body"].iter_chunks()
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+        _LOGGER.debug("Uploading backup with ID %s", backup.backup_id)
+
+        key = _get_key(backup)
+        metadata = {
+            "metadata_version": METADATA_VERSION,
+            "backup_metadata": backup.as_dict(),
+        }
+
+        multipart_upload = await self._client.create_multipart_upload(
+            Bucket=self._bucket,
+            Key=key,
+            Metadata=_serialize(metadata),
+        )
+
+        upload_id = multipart_upload["UploadId"]
+        parts = []
+        part_number = 1
+
+        stream = await open_stream()
+        async for chunk in stream:
+            part = await self._client.upload_part(
+                Bucket=self._bucket,
+                Key=key,
+                PartNumber=part_number,
+                UploadId=upload_id,
+                Body=chunk,
+            )
+            parts.append({"PartNumber": part_number, "ETag": part["ETag"]})
+            part_number += 1
+
+        await self._client.complete_multipart_upload(
+            Bucket=self._bucket,
+            Key=key,
+            UploadId=upload_id,
+            MultipartUpload={"Parts": parts},
+        )
+
+    async def async_delete_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> None:
+        """Delete a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        """
+        _LOGGER.debug("Deleting backup with ID %s", backup_id)
+        try:
+            _, obj = await self._get_backup(backup_id)
+        except ValueError:
+            return
```

Comment:
```suggestion
            raise BackupNotFound 
```
if we don't find something here, we actually want to raise the error